### PR TITLE
Installer Overhaul

### LINKED ISSUE #1683: New installer
The current installer works fine but is a bit clunky, unclear, and outdated in some ways. 

A replacement with improved UX and not reliant on jQuery (#907) should be implemented.

### PR DESCRIPTION
**Pull request by Ron-Zarbyte**
Closes #1683 

Overhauls and optimizes the installer as far as I can go without rewriting the whole thing. I am happy enough with the end result to not think a full rewrite is justified.

My goal was a one-to-one replacement of the existing UI with a few enhancements combined with a theme as lightweight as possible. Some of the UI is not perfect and could perhaps be improved by someone who enjoys frontend more, however I made sure no parts of the UI are _worse_ than the old installer. Everything is the same or an improvement, and it is now much more easily readable and expandable.

Please review thoroughly, attempt to break it, and do not reserve feedback. The installer is the first thing new users see in the software so UX is important. I will not be offended.

# Changes

- Implemented [Pico.css](https://picocss.com/) for the new frontend template and removed all old assets. The theme will automatically adapt to the system's light/dark mode preference.
- jQuery has been fully removed to stay inline with #907. The remaining javascript is vanilla.
- Added `is subfolder` check. This blocks installation if we detect the user is trying to install in a subfolder. (trying to put a stop to all the discord questions... 😵)
- Added `already installed` check. The installer is also now protected from tampered install requests. This prevents the installer from being able to run if already installed, even if the `install` folder is left on the server, and if a user tries to `POST` directly to the install trigger endpoint. With this change I feel it is no longer a security issue to leave the installer on the webserver, however there is also not really a reason to adjust that behavior. It will be safer to keep removing it. This was just for my peace of mind.
- Added Requirements blocker for the setup form. If any system requirements fail, a new message is shown, and the setup form is hidden.
- Too many readability and flow improvements to list. Removed unnecessary functions, removed duplicate function calls, properly used local class variables rather than passing references around everywhere...
- Removed the `Agree` checkbox. It was checked by default anyway and I don't really think it is necessary for FOSSBilling.
- Removed and avoided any javascript alerts. The old confirmation alert was not styled and a little unnecessary anyway.
- Updated the Install Complete page to give the user references of the exact commands to run.

I am likely missing something but those are the major points.

# Screenshots

image
image
image
image
image

### CODE REVIEW

**Comment by John-S4** [comment]
Ooh, exciting :)

I haven't had a chance to look much yet, but one initial thought just from the screenshots is that we should add something in documentation about the lack of subfolder support and reference that in the failure message for subfolder installations rather than the GH issue.

**Comment by jaapmarcus** [comment]
> Ooh, exciting :)
> 
> I haven't had a chance to look much yet, but one initial thought just from the screenshots is that we should add something in documentation about the lack of subfolder support and reference that in the failure message for subfolder installations rather than the GH issue.

Makes sense to include the information in our docs instead of a Github issue.

**Comment by BelleNottelling** [comment]
Thanks for the work @Ron-Zarbyte, it's greatly appreciated.

I've just spent some time improving the actual layout and visuals of the new installer, it is IMO better than what you had and significantly better than the old version.
The biggest changes are:
1. Some of the more requirements checks are hidden unless there is an issue with them, as a lot of them aren't going to come up or matter for 95% of users.
2. Windows will now create a warning
3. The page layout has been completely changed to use accordions to separate each item by "steps" & JS is used to prevent jumping to the next step before the previous ones are competed.
4. I've reduced the quantity of large spacing as it was pretty significant in your version.
5. I have moved it to using locale CSS rather than from over a CDN. We don't really like using CDNs for anything wherever possible.

## New design I made
Animation

image

Overall, I think the version I pushed is quite a bit better than either option we had, but I'm happy to get feedback on it.
Probably all that I'd change is the JS as I'm assuming there is a better way to do what I did, but it's at least functional and not *too* horrible.

I'd also like to hear what people think about the removal of the "I agree" checkbox. I am in favor of keeping it and having people acknowledge they know what FOSSBilling is licensed under.

**Comment by BelleNottelling** [comment]
> that we should add something in documentation about the lack of subfolder support and reference that in the failure message for subfolder installations rather than the GH issue.

I don't really see the point of listing anything at all. Unless we intend to write out a technical explanation, there's not much to add outside of "it doesn't work right now"

**Comment by evrifaessa** [comment]
@Ron-Zarbyte and @BelleNottelling, thanks for the awesome work.

I believe that Belle's changes are justified. It's also lovely to see the steps kept while also getting rid of jQuery. I think it's easier to follow for a user and easier for us to put value validation in between compared to a long page.

About the "I agree" checkbox, I'm in favor of keeping it.

**Comment by evrifaessa** [comment]
My bad - accidentally closed the issue.

Another thing is, there should be a consistency among the casing of the titles. "Check System Requirements" and "Enter database details" don't match. I'm in favor of the latter.

**Comment by John-S4** [comment]
> I don't really see the point of listing anything at all. Unless we intend to write out a technical explanation, there's not much to add outside of "it doesn't work right now"

I disagree with this, I wrote something together with some other minor docs changes/additions but I'm not near computer right now so will PR them later.

**Comment by John-S4** [comment]
> Another thing is, there should be a consistency among the casing of the titles. "Check System Requirements" and "Enter database details" don't match. I'm in favor of the latter.

Agree - sentence case.

**Comment by evrifaessa** [comment]
Just tried on a live install. The installation goes well and the database is populated. However, I'm redirected to `/install/install.php?a=install` which is blank.

**Comment by BelleNottelling** [comment]
> I disagree with this, I wrote something together with some other minor docs changes/additions but I'm not near computer right now so will PR them later.

Alright, I'll wait to see what you came up with. By the way, what are your thoughts on having a "I agree" checkbox in the installer for the FOSSBilling license? @John-S4 & @jaapmarcus.

> Just tried on a live install. The installation goes well and the database is populated. However, I'm redirected to /install/install.php?a=install which is blank.

I saw the same behavior, but haven't yet worked to identify the cause.

**Comment by Ron-Zarbyte** [comment]
The UI adjustments are looking really good! I tried to make it a "one click" installer, but the page just gets a bit too long. The newer stepper UI flows better.

The only other adjustment I'd bother to mention is maybe side by side Back/Next buttons instead of stacked, with the Back button using secondary color for some visual separation. Perhaps some type of color separation between the Installer button as well, or on the last step replace the Next button with Install instead of three buttons with one disabled. But that is just me trying to find something to give feedback about.

I will look at the installer white page when I get back at my desk in approximately 5-6 hours from now. Probably easier for me to debug/handle that versus someone else reading installer code. I used docker for main testing so it is more than likely just an environment difference I need to factor in. I will put up a remote installation and find the bug.

**Comment by evrifaessa** [comment]
> The only other adjustment I'd bother to mention is maybe side by side Back/Next buttons instead of stacked, with the Back button using secondary color for some visual separation. Perhaps some type of color separation between the Installer button as well, or on the last step replace the Next button with Install instead of three buttons with one disabled. But that is just me trying to find something to give feedback about.

Nice suggestion :)

**Comment by BelleNottelling** [comment]
Fixed the blank screen issue - it appeared to be as a result of the code deleting the install directory and then attempting to render the page and sending it to the browser. Reversing this has resolved the problem for me

**Comment by evrifaessa** [comment]
> Reversing this has resolved the problem for me

Can confirm. Thanks for the patch :)

**Comment by John-S4** [comment]
> The only other adjustment I'd bother to mention is maybe side by side Back/Next buttons instead of stacked, with the Back button using secondary color for some visual separation. Perhaps some type of color separation between the Installer button as well, or on the last step replace the Next button with Install instead of three buttons with one disabled. But that is just me trying to find something to give feedback about.

100% agree, I looked at those huge page width stacked buttons and just thought 'please no' but the version in @BelleNottelling's later video with smaller buttons looks much nicer.

**Comment by John-S4** [comment]
> By the way, what are your thoughts on having a "I agree" checkbox in the installer for the FOSSBilling license? @John-S4 & @jaapmarcus.

I think that it should be there, but in light of some of the discussion around legal liability stemming from Issue 1678 I actually think that the checkbox should be not just for the Apache license but also agreeing that 'the software is provided as is, and that it is the responsibility of the user to ensure that they use it in accordance with local laws in their state' or something along those lines.

**Comment by Anuril** [comment]
> > By the way, what are your thoughts on having a "I agree" checkbox in the installer for the FOSSBilling license? @John-S4 & @jaapmarcus.
> 
> I think that it should be there, but in light of some of the discussion around legal liability stemming from #1678 I actually think that the checkbox should be not just for the Apache license but also agreeing that 'the software is provided as is, and that it is the responsibility of the user to ensure that they use it in accordance with local laws in their state' or something along those lines.

I do not see it as necessary to display the apache license, absolutely nobody reads licenses and making it clear that the project is licensed under the Apache 2 license and that there is no warranty whatsoever should be enough for anyone to be informed.

I have updated the original post with more clear communication. 

Also - if I am the reason that a license field is added to any software, that'd be weighing on my conscience heavily.

**Comment by John-S4** [comment]
> Also - if I am the reason that a license field is added to any software, that'd be weighing on my conscience heavily. (For I hate it with a passion.)

The checkbox for the license has always been there, it was only removed by this PR.

File: src/install/assets/install.html.twig [hunk 1]
```diff
@@ -1,396 +1,290 @@
...
+
+                    <fieldset>
+                        <label for="database_hostname">Hostname <span class="red">*</span></label>
+                        <input type="text" id="database_hostname" name="database_hostname" value="{{ database_hostname|default('localhost') }}" placeholder="localhost, mysql, db.domain.com, 12.34.56.78 ..." required>
+                        <small>Usually <strong>localhost</strong>. For Docker, use <strong>mysql</strong>.</small>
+    
+                        <label for="database_port">Port <span class="red">*</span></label>
+                        <input type="number" id="database_port" name="database_port" value="{{ database_port|default('3306') }}" placeholder="Usually 3306." required>
+    
+                        <label for="database_name">Name <span class="red">*</span></label>
```
Comment (Anuril) [inline src/install/assets/install.html.twig:118]: I would suggest using "Database Name" as the label here, as there is another Name field for the Admin later.

**Comment by Anuril** [comment]
Except the Label, LGTM

File: src/install/assets/install.html.twig (hunk 1 above)
Comment (John-S4) [reply src/install/assets/install.html.twig:118]: Yep, I would agree that makes sense.

**Comment by John-S4** [comment]
> we should add something in documentation about the lack of subfolder support and reference that in the failure message for subfolder installations rather than the GH issue.

https://fossbilling.org/docs/troubleshooting#subfolder-installations

**Comment by BelleNottelling** [comment]
> > we should add something in documentation about the lack of subfolder support and reference that in the failure message for subfolder installations rather than the GH issue.
> 
> https://fossbilling.org/docs/troubleshooting#subfolder-installations

So we now have two separate items in the troubleshooting guide about sub-folder installations.
One that simply references sub-folders in it's title (and is wrongly placed under the 404 errors) and one that's listed under the actual error people will receive when trying to use a sub-folder:
https://fossbilling.org/docs/troubleshooting#subfolder-installations
-and-
https://fossbilling.org/docs/troubleshooting#unknown-encore_entry_link_tags-function

**Comment by John-S4** [comment]
@BelleNottelling yeah I only noticed the other one at the bottom after I added one.

However it is under the 404 error be Use the last screenshot people sent of it not working under subfokders were 404 pages.

I will combine and fix ehne I'm at my desk though.

**Comment by BelleNottelling** [comment]
> However it is under the 404 error because the last screenshots people sent of it not working under subfolders were 404 pages.

The 404 errors shouldn't really be because of it being under a sub-folder, that's generally because they don't have any config running for FOSSBilling on their web-server, resulting in the virtual paths not being forwarded to `index.php` to be handled by FOSSBilling.

The actual technical reason why sub-folders aren't supported is because there's some code in the back-end that enables encore when it detects that the current theme needs to have it. When people put FOSSBilling under a sub-folder that code fails to correctly detect that the current route is to the administrator panel, resulting in it not enabling encore for the administrator theme. Then when the admin panel's default template tries to load in it's assets using encore, they are hit with a `Unknown "encore_entry_link_tags" function.` error from Twig

**Comment by John-S4** [comment]
> The actual technical reason why sub-folders aren't supported is because...

Yes, I get your point, but if they are getting 404s first, which many seem to be, then I think it makes most sense to just say straight off this isn't going to work in a sub-folder rather than have them try to solve other issues first and then end up at the point that it isn't going to work anyway because it's a sub-folder installation. 

I will combine both texts that are there, leaving a note in the 404 section and directing to the other section. 

In any case, the point for the purposes of this PR is that we should link to the troubleshooting page in docs from the error in the installer and not to the GitHub issue.

**Comment by Ron-Zarbyte** [comment]
I applied all of the text suggestions and adjusted the subfolder reference link from the issue to `https://fossbilling.org/docs/troubleshooting#subfolder-installations` instead.

I think that does it for the outstanding changes but let me know if we need anything else on my side.

Finally had a chance to look at the stepper UI in browser. It is looking great! 😁

> The actual technical reason why sub-folders aren't supported is ...

Interesting, and good to know. Thanks for that info.

**Comment by BelleNottelling** [comment]
> I think that does it for the outstanding changes but let me know if we need anything else on my side.

I believe the team agreed that we'd want to have a have a checkbox for people to verify they agree to the FOSSBilling license, acknowledge that they understand the software is provided as-is, and accept that it is their responsibility to verify they are using it in accordance with their local laws.

Other than that, I think everything has been addressed.

> Interesting, and good to know. Thanks for that info.

The relevant code for this is [here](https://github.com/FOSSBilling/FOSSBilling/blob/main/src/modules/Theme/Service.php#L425) if you're at all curious on it. I'm sure it's a very fixable problem, we just aren't super familiar with the routing system as it's (to the best of my knowledge) custom and originates from BoxBilling, so it's creation predates the FOSSBilling team and we are learning it as we go along.

We also just haven't tried to fix it yet because.. well, people should really just be separating things over domains & sub-domains rather than folders.

File: src/install/assets/install.html.twig [hunk 2]
```diff
@@ -1,396 +1,290 @@
...
+                    <fieldset>
+                        <label for="currency_code">Code <span class="red">*</span></label>
+                        <input type="text" id="currency_code" name="currency_code" value="{{ request.currency_code|default(currency_code) }}" required>
+    
+                        <label for="currency_title">Title <span class="red">*</span></label>
+                        <input type="text" id="currency_title" name="currency_title" value="{{ request.currency_title|default(currency_title) }}" required>
+    
+                        <label for="currency_format">Price format <span class="red">*</span></label>
+                        <input type="text" id="currency_format" name="currency_format" value="{{ request.currency_format|default(currency_format) }}" required>
+                    </fieldset>
```
[the remark this answers is not in the document]
Comment (Successfulsebunya) [reply src/install/assets/install.html.twig:160]: How about you skip the checkbox and just put buttons to continue or cancel????

I agree or I don't agree

"I agree" proceeds to take the user to next page

"I don't agree" closes the page

**Comment by Ron-Zarbyte** [comment]
I completely overlooked the agreement checkbox, sorry about that. I've added it in now with the suggested location & text.

chrome_yqGAFhmFkD


I will take a look at the subfolder issue for my own curiosity. But I agree with it is best to not be concerned with it, especially with the subfolder blocker on the installer now. People won't be confused by it anymore.

**Comment by jaapmarcus** [comment]
Shouldn't not be the first step to agree with the license  in stead the last step before installing?

**Comment by John-S4** [comment]
> Shouldn't not be the first step to agree with the license in stead the last step before installing?

Logically yes, probably. But when was the last time you used a checkout or signup or download form that didn't put it last, it's just the convention for some reason.

**Comment by jaapmarcus** [comment]
> > Shouldn't not be the first step to agree with the license in stead the last step before installing?
> 
> Logically yes, probably. But when was the last time you used a checkout or signup or download form that didn't put it last, it's just the online convention for some reason.

Windows displays first the license and then allows you to install...

For "shopping" it mostly does when you create your user account and before the payment step...

**Comment by Ron-Zarbyte** [comment]
When whoever overhauled the original UI with the stepper, the form was placed only within steps 1 ~ 4. The UI would need to be adjusted again to allow the checkbox to be placed outside of the stepper while allowing the stepper validation to still work.

For example... the checkbox couldn't easily be moved (copy/pasted) right now to the requirements area before the setup form, because the form tags start and end within the stepper, which also has some javascript requirements attached with it now. You could perhaps add an additional submission form to the page outside of the stepper that requires checkbox validation before allowing you to see the setup form.

I don't like the latter as I feel two forms on the same page is not necessary.

I also don't really care to adjust placement myself considering the extra complexity now with the stepper UI in place. Someone who favors working with javascript more than me can look at that one if desired.

Otherwise you're limited to placing the checbox somewhere between the current form tags. Feel free to push an adjustment if something else works better.

I am still of the opinion that this checkbox is not necessary at all from a legal perspective. The Apache license already covers the end user. The Apache 2.0 license already removes all liability from open source software maintainers in the event software is exploited due to bad design. At the end of the day this checkbox is just a formality that doesn't even mean anything because it is already covered by the Apache license.

I'd much rather see the Agreement checkbox entirely removed for this reason. It is also why I removed it in the original PR. I understand there was some legal precaution going on, but the license itself directly states we are not liable. Its part of the purpose of using this license.

> 8. Limitation of Liability. In no event and under no legal theory, whether in tort (including negligence), contract, or otherwise, unless required by applicable law (such as deliberate and grossly negligent acts) or agreed to in writing, shall any Contributor be liable to You for damages, including any direct, indirect, special, incidental, or consequential damages of any character arising as a result of this License or out of the use or inability to use the Work (including but not limited to damages for loss of goodwill, work stoppage, computer failure or malfunction, or any and all other commercial damages or losses), even if such Contributor has been advised of the possibility of such damages.

Ref: https://www.apache.org/licenses/LICENSE-2.0

So +1 to just removing the checkbox.

**Comment by BelleNottelling** [comment]
> When whoever overhauled the original UI with the stepper, the form was placed only within steps 1 ~ 4. The UI would need to be adjusted again to allow the checkbox to be placed outside of the stepper while allowing the stepper validation to still work.

The work to do this is pretty minimal. I can add it either as the first or last step depending on what @John-S4, @jaapmarcus, @evrifaessa, and @Anuril think.

> I am still of the opinion that this checkbox is not necessary at all from a legal perspective. The Apache license already covers the end user's acceptance of their own liability. The Apache 2.0 license already removes all liability from open source software maintainers in the event software is exploited due to bad design. At the end of the day this checkbox is just a formality that doesn't even mean anything because it is already covered by the Apache license.

I get your perspective, however at the end of the day, people don't read things unless they are forced to do so. This has been made very clear to us many times over. If you follow the Discord at all, you will see a lot of people asking the same questions again and again. Questions that are answered in our FAQ and troubleshooting documentation and even though we even go as far as to link to this information in our discord rules and ask that people read them before asking questions, we still see the same things again and again. People don't read.

Likewise, about a week ago I added a note to the downloads page that clearly states FOSSBilling does not work under a sub-folder and we still see people trying to do so. People don't read. 

I also fully expect many people to just block out the top portion of the installer from their brain and not even realize it is there because it's not directly linked to their ability to complete and installation of FOSSBilling.

While such a checkbox might not legally be necessary, forcing someone to click it should at least help reduce the likelihood that they install it without understanding what the legal conditions of using FOSSBilling are and in the event that someone still doesn't, we can say it is 100% their own fault because we made them click that checkbox.

**Comment by evrifaessa** [comment]
I already know about the Apache 2.0 covering it.

Like Belle said, people don't read licenses and it wouldn't hurt to give them a heads up one more time.

**Comment by Successfulsebunya** [comment]
> I completely overlooked the agreement checkbox, sorry about that. I've added it in now with the suggested location & text.
> 
> chrome_yqGAFhmFkD
> 
> 
> I will take a look at the subfolder issue for my own curiosity. But I agree with it is best to not be concerned with it, especially with the subfolder blocker on the installer now. People won't be confused by it anymore.
> 
> Edit:
> 
> @Successfulsebunya while I do not really think either one is necessary personally, I would at least agree with the checkbox over a button like `"Agree & Continue"`. The explicit agreement is better than an implicit agreement confirmation that could easily be confused as just another `"Next"` button.

You are right. Feels right too.

**Review by BelleNottelling — APPROVED** [review]

**Review by Anuril — APPROVED** [review]
LGTM :) Let's merge this!

**Pull request merged**